src/v/cluster/topics_frontend.cc

```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
     if (!units) {
         co_return make_error_result(assignable_config.cfg.tp_ns, units.error());
     }
+
+    // Enforce cluster-wide topic count limit only on user created topics.
+    auto max_user_topics_opt = _max_user_topics();
+    if (max_user_topics_opt.has_value()) {
+        // This is intended to approximate the number of topics that are
+        // user-defined and may return a count slightly smaller than the actual
+        // amount.
+        auto user_topic_count = std::max(
+                                  _topics.local().all_topics_count(),
+                                  model::non_user_topics.size())
+                                - model::non_user_topics.size();
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
```

Comment:
Seems the `is_user_topic` check should just go on the outer if?

Reply:
I mainly left it as the final check as its the most expensive. It may hurt readability a bit, however, I wouldn't expect this limit to be hit in most cases and therefore wanted its addition to have the least overhead possible.